### ISSUE TITLE
SNOW-636720: Panic due to slice bounds out of range in paddingTrim

### ISSUE THREAD

**Issue by Thomasdezeeuw** [snowflakedb/gosnowflake#631]
### Issue description

Panic due to slice bounds out of range, see backtrace below.

### Example code

I can't share at the moment.

### Error log

Trace:

```
panic: runtime error: slice bounds out of range [:-150]

goroutine 114 [running]:
github.com/snowflakedb/gosnowflake.paddingTrim(...)
        /Users/thomas/go/pkg/mod/github.com/snowflakedb/gosnowflake@v1.6.11/encrypt_util.go:277
github.com/snowflakedb/gosnowflake.decryptFile(0x14000311f50, 0x14000314c00?, {0x140005da240, 0x2b}, 0x0, {0x1400031a000, 0x3a})
        /Users/thomas/go/pkg/mod/github.com/snowflakedb/gosnowflake@v1.6.11/encrypt_util.go:206 +0x518
github.com/snowflakedb/gosnowflake.(*remoteStorageUtil).downloadOneFile(0x0?, 0x14000314c00)
        /Users/thomas/go/pkg/mod/github.com/snowflakedb/gosnowflake@v1.6.11/storage_client.go:213 +0x484
github.com/snowflakedb/gosnowflake.(*snowflakeFileTransferAgent).downloadOneFile(0x1400051cf00, 0x14000314c00)
        /Users/thomas/go/pkg/mod/github.com/snowflakedb/gosnowflake@v1.6.11/file_transfer_agent.go:986 +0x17c
github.com/snowflakedb/gosnowflake.(*snowflakeFileTransferAgent).downloadFilesParallel.func1(0x1, 0x0?)
        /Users/thomas/go/pkg/mod/github.com/snowflakedb/gosnowflake@v1.6.11/file_transfer_agent.go:886 +0x7c
created by github.com/snowflakedb/gosnowflake.(*snowflakeFileTransferAgent).downloadFilesParallel
        /Users/thomas/go/pkg/mod/github.com/snowflakedb/gosnowflake@v1.6.11/file_transfer_agent.go:884 +0x264
```

### Configuration

*Driver version (or git SHA):* gosnowflake@v1.6.11

*Go version:* go version go1.18.4 darwin/arm64

**Comment by Thomasdezeeuw** [comment]
Any update on this? This is blocking our support for Snowflake.

**Comment by Thomasdezeeuw** [comment]
It's been two months now... we're trying to support Snowflake, but without a response this makes it quite difficult.

**Comment by ColmanHumphrey** [comment]
I don't have a repro just yet, but for a quick check:

In https://github.com/snowflakedb/gosnowflake/blob/82aecfae87c571aa7dfe1c7542e1e48517e79ec8/encrypt_util.go#L287-L290, there are no checks on the unpadding. Compare this to say https://github.com/go-web/tokenizer/blob/f87b3c7a4e753ae3ad7f95b270cfeae509083e99/pkcs7.go#L52-L62, where nearly the same thing is happening. There, there's an explicit check if:
- the padding is 0, or too big
- any of the padded elements are different

I'm guessing this code should have failed the first check, that the padding was too big, and thrown a padding error back.

**Comment by sfc-gh-dszmolka** [comment]
Hi,
The driver dev team made a new branch of the driver which is hoped to resolve issues #630, #631, #632.
https://github.com/snowflakedb/gosnowflake/tree/fixCipherPanicIssues

Would it be please possible @thomas to test it and see if it really does ? Would appreciate any feedbacks.

**Comment by Thomasdezeeuw** [comment]
@sfc-gh-dszmolka I can confirm that the tree fixes this issue. Thanks @sfc-gh-ext-simba-lb!

**Comment by sfc-gh-dszmolka** [comment]
fixed by #687 earlier
